File: src/leetcode/linkedlist/ReverseKGroup.java
package leetcode.linkedlist;

/**
 * @author chengzw
 * @description K个一组翻转链表，https://leetcode-cn.com/problems/reverse-nodes-in-k-group/
 * @since 2021/8/29
 */
public class ReverseKGroup {

    public static ListNode solution1(ListNode head, int k) {

        //虚拟节点
        ListNode dummyNode = new ListNode();
        ListNode tail = dummyNode;

        //p是原链表遍历指针
        ListNode p = head;
        while (p != null) {
            //计数，达到k个，反转链表
            int count = 0;
            //q遍历k个节点的链表
            ListNode q = p;
            while (q != null) {
                count++;
                if (count == k) {
                    break;
                }
                q = q.next;
            }

            //如果剩下的不到k个，是 q!=null 退出循环的
            //把剩下的添加到新链表后，不反转
            if (q == null) {
                tail.next = p;
                return dummyNode.next;
            } else {  //反转k个节点的链表
                //p传入reverseList以后回变化，先用tmp保存
                ListNode tmp = q.next;
                ListNode[] nodes = reverseList(p, q);
                tail.next = nodes[0];
                tail = nodes[1];
                p = tmp;
            }
        }
        return dummyNode.next;
    }


    //反转链表
    //传入k个节点的链表的头节点和尾节点
    //返回反转的链表头节点和尾节点
    public static ListNode[] reverseList(ListNode head, ListNode tail) {


        //p是k个长度的链表的遍历指针
        ListNode p = head;
        //头节点
        ListNode newHead = null;
        //反转链表
        while (p != tail) {
            ListNode tmp = p.next;
            p.next = newHead;
            newHead = p;
            p = tmp;
        }

        tail.next = newHead;
        return new ListNode[]{tail, head};
    }

    public static void main(String[] args) {
        ListNode result1 = solution1(ListNode.createNode(new int[]{1, 2, 3, 4, 5}), 2);
        System.out.println(ListNode.printNode(result1));
    }
}

File: go/leetcode/linkedlist/reverse_k_group.go
package linkedlist

/**
 * @description 25. K 个一组翻转链表
 * @author chengzw
 * @since 2022/5/23
 * @link https://leetcode.cn/problems/reverse-nodes-in-k-group/
 */

/**
给你链表的头节点 head， 每 k 个节点一组进行翻转， 请你返回修改后的链表。
k 是一个正整数， 它的值小于或等于链表的长度。 如果节点总数不是 k 的整数倍， 那么请将最后剩余的节点保持原有顺序。
你不能只是单纯的改变节点内部的值， 而是需要实际进行节点交换。

示例 1：
输入：head = [1,2,3,4,5], k = 2
输出：[2,1,4,3,5]

示例 2：
输入：head = [1,2,3,4,5], k = 3
输出：[3,2,1,4,5]
*/

/**
思路：
1.两个指针，p 遍历原链表，q 遍历 k 个一组链表
2.如果是 q == nil 退出的，说明剩下的几点不到 k 个，直接接到新链表后面，不反转。否则先反转后再接到新链表后面。

时间复杂度: O(n)
空间复杂度: O(1)
*/
func reverseKGroup(head *ListNode, k int) *ListNode {
	// 新链表
	newHead := new(ListNode)
	tail := newHead
	// 遍历原链表
	p := head
	for p != nil {
		// 遍历 k 个一组链表
		q := p
		count := 0
		for q != nil {
			count++
			if count == k {
				break
			}
			q = q.Next
		}

		// 处理 k 个链表
		// 如果是 q == nil 退出的，说明剩下的几点不到 k 个，直接接到新链表后面，不反转
		if q == nil {
			tail.Next = p
			return newHead.Next
		} else { // 先反转后再接到新链表后面
			tmp := q.Next
			reverseHead, reverseTail := reverseKList(p, q)
			tail.Next = reverseHead
			tail = reverseTail
			p = tmp
		}
	}
	return newHead.Next
}

// 反转链表
func reverseKList(head, tail *ListNode) (*ListNode, *ListNode) {
	var k *ListNode
	p := head
	for p != tail {
		tmp := p.Next
		p.Next = k
		k = p
		p = tmp
	}
	tail.Next = k
	return tail, head
}
